Storage/SQL/SQLiteHistory.swift

```diff
@@ -137,39 +149,65 @@ extension SQLiteHistory: BrowserHistory {
+
+            // Insert instead.
+            return self.insertSite(site, atTime: now, isLocal: true, withConnection: conn)
+        }
+
+        return failOrSucceed(error, "Record site")
+    }
+
+    func updateSite(site: Site, atTime time: NSNumber, withConnection conn: SQLiteDBConnection) -> Int {
+        // We know we're adding a new visit, so we'll need to upload this record.
+        // If we ever switch to per-visit change flags, this should turn into a CASE statement like
+        //   CASE WHEN title IS ? THEN max(should_upload, 1) ELSE should_upload END
+        // so that we don't flag this as changed unless the title changed.
+        //
+        // Note that we will never match against a deleted item, because deleted items have no URL,
+        // so we don't need to unset is_deleted here.
+        if let domain = site.domain {
+            let update = "UPDATE \(TableHistory) SET title = ?, local_modified = ?, should_upload = 1, domain_id = (SELECT id FROM \(TableDomains) where domain = ?) WHERE url = ?"
+            let updateArgs: Args? = [site.title, time, domain, site.url]
             if LogPII {
                 log.debug("Setting title to \(site.title) for URL \(site.url)")
             }
-            error = conn.executeChange(update, withArgs: updateArgs)
+            var error = conn.executeChange(update, withArgs: updateArgs)
             if error != nil {
-                log.warning("Update failed with \(err?.localizedDescription)")
+                log.warning("Update failed with \(error?.localizedDescription)")
                 return 0
             }
-            if conn.numberOfRowsModified > 0 {
-                return conn.numberOfRowsModified
+            return conn.numberOfRowsModified
+        }
+        return 0
+    }
+
+
+    private func insertSite(site: Site, atTime time: NSNumber, isLocal: Bool, withConnection conn: SQLiteDBConnection) -> Int {
+        var error: NSError? = nil
+
+        if let domain = site.domain {
+            if let error = conn.executeChange("INSERT OR IGNORE INTO \(TableDomains) (domain) VALUES (?)", withArgs: [domain]) {
+                log.warning("Domain Insert failed with \(error.localizedDescription)")
+                return 0
             }
 
-            // Insert instead.
-            let insert = "INSERT INTO \(TableHistory) (guid, url, title, local_modified, is_deleted, should_upload) VALUES (?, ?, ?, ?, 0, 1)"
-            let insertArgs: Args? = [Bytes.generateGUID(), site.url, site.title, now]
-            error = conn.executeChange(insert, withArgs: insertArgs)
-            if error != nil {
-                log.warning("Insert failed with \(err?.localizedDescription)")
+            let insert = "INSERT INTO \(TableHistory) " +
+                "(guid, url, title, " + (isLocal ? "local_modified" : "server_modified") + ", is_deleted, should_upload, domain_id) " +
+                "SELECT ?, ?, ?, ?, 0, 1, id FROM \(TableDomains) where domain = ?"
```

Comment:
This will do weird things if the domain table isn't in the state we expect. Also I think you're setting `should_upload` to 1, which will be incorrect for the Sync case. These are basically two totally different code paths…